### PR TITLE
cilium node chain refactor

### PR DESCRIPTION
**Pull request by bimmlerd**
The intent of this PR is to remove the CiliumNodeChain.

Review commit by commit.

bgp/speaker: use Resource[T] instead of NodeChains
    
    As part of a refactoring effort to remove the CiliumNodeChain, remove
    the dependency of the MetalLBSpeaker on it, by directly using the
    LocalCiliumNodeResource.
    
    Note that the speaker always discarded events not pertaining to the
    local Node, hence no functional change results from the fact that the
    CiliumNodeChain proagated events from all cluster CiliumNodes.
    
    As an added benefit, the CiliumNode events now don't stop coming once we
    connect to KVStore, since the resource doesn't stop.
   

ipam: multipool: use resource for node events

```
As part of a refactoring effort to remove the CiliumNodeChain, remove
the multipool IPAM dependency on it.

Since IPAM isn't yet part of the hive, we need to pass the resource down
to IPAM through the daemon, in `startIPAM`. Once IPAM code is
modularised, this is no longer necessary.

In order to keep the tests largely unchanged, we reimplement the
fakeNodeK8sAPI testing infrastructure in terms of the resource API. This
leads to code duplication, but a commit later in this series removes the
duplication again, by also converting the other usages to the resource style.
```

watchers: remove ciliumnodechain

    Previous commits in the series have removed all subscribers to the
    CiliumNodeChain by moving them over to use the LocalCiliumNodeResource.
    This commit reaps the benefits by deleting all of the now unused code.

### CODE REVIEW

File: pkg/k8s/watchers/subscriber/ciliumnode.go
Comment (joamaki) [inline pkg/k8s/watchers/subscriber/ciliumnode.go]: Doesn't this only return the error from last subscriber since we overwrite `err`?
`err = errors.Join(err, s.OnAddCiliumNode(node))` would probably not work since we get nested `joinError`s. Perhaps just do the same as the original except use `errors.Join(errs...)` instead of `fmt.Errorf`.

File: pkg/k8s/watchers/subscriber/ciliumnode.go
Comment (joamaki) [inline pkg/k8s/watchers/subscriber/ciliumnode.go]: Same issue as above.

File: pkg/k8s/watchers/subscriber/ciliumnode.go
Comment (joamaki) [inline pkg/k8s/watchers/subscriber/ciliumnode.go]: Same issue as above.

File: pkg/k8s/watchers/subscriber/ciliumnode.go
Comment (joamaki) [inline pkg/k8s/watchers/subscriber/ciliumnode.go]: Should we just rename this to `LocalCiliumNodeChain` so there won't be confusion in case refactoring takes a bit longer?

**Review by joamaki — CHANGES_REQUESTED** [review]
~~The approach LGTM. Fix the error joining :)~~

The approach is flawed as we're now retrying an event that single subscriber fails to process on all of the subscribers in CiliumNodeChain. I'm not sure they're idempotent nor would it be the behavior we want. See my last comment.

File: pkg/k8s/watchers/subscriber/ciliumnode.go
Comment (joamaki) [inline pkg/k8s/watchers/subscriber/ciliumnode.go]: This will now retry the event for all subscribers even if only one of them failed. I'm not sure we can rely on them being idempotent. It might be better to keep the original semantics of not doing much with the error (it was just marking up the EventProcessed metric). Though then we won't have `Resource[T]` marking up the metric and will lose metric coverage :(

Take a look at the subscribers and see what options we have.

Comment (bimmlerd) [reply pkg/k8s/watchers/subscriber/ciliumnode.go]: Thanks for this catch. Rewrote the PR to drop CiliumNodeChain completely

File: daemon/cmd/ipam.go
Comment (gandro) [inline daemon/cmd/ipam.go]: nit: Maybe we should call this `node`

File: pkg/ipam/multipool.go
Comment (gandro) [inline pkg/ipam/multipool.go]: Question: Does the resource watcher issue an event if the `resourceVersion` has changed, but the object is otherwise semantically equivalent? The current IPAM code relies on this (though this can be changed if the nodeStore gives us the most recent version)

**Review by gandro — APPROVED** [review]
Thanks so much for tackling this. One question around upsert events, otherwise this looks excellent to me

**Review by rastislavs — APPROVED** [review]
LGTM for BGP

File: pkg/ipam/multipool.go
Comment (bimmlerd) [reply pkg/ipam/multipool.go]: Yes.

Resource[T] doesn't do any deep equal stuff. The only deduplication that happens is if an event hasn't been observed, and a newer event regarding the same object comes in, you only get one event. This isn't resource code, this is just the k8s workqueue implementation. Since `Resource[T].Events()` fetches the newest state for an object (by key) from the store for each event it delivers anyway, this just means you don't get two identical events.

File: pkg/bgp/speaker/speaker.go
Comment (joamaki) [inline pkg/bgp/speaker/speaker.go]: shouldn't this use the context passed to `New`? Otherwise only way to stop here is to stop the resource.

File: pkg/ipam/multipool.go
Comment (joamaki) [inline pkg/ipam/multipool.go]: Do we have a proper context to use here that is cancelled when IPAM stops? This might be dangerous as Resource[T] might be emitting events while systems that this is calling into are being stopped.

File: pkg/ipam/clusterpool.go
Comment (joamaki) [inline pkg/ipam/clusterpool.go]: Please address the FIXME

File: pkg/ipam/clusterpool.go
Comment (joamaki) [inline pkg/ipam/clusterpool.go]: Same here as in previous comments: use a proper context that is cancelled when this sub-system is stopping. Or if there is no such thing, validate that it's fine that this only stops when the `Resource[T]` stops and it's fine handling events from it while everything is being stopped & add comment to explain this.

**Review by joamaki — CHANGES_REQUESTED** [review]

File: pkg/bgp/speaker/speaker.go
Comment (danehans) [inline pkg/bgp/speaker/speaker.go:21]: nit: For consistency, consider using `_` for named imports, e.g. `agent_k8s`.

File: daemon/cmd/ipam.go
Comment (danehans) [inline daemon/cmd/ipam.go:17]: nit: For consistency, named imports should use underscores.

File: pkg/ipam/multipool.go
Comment (danehans) [inline pkg/ipam/multipool.go:18]: Same feedback as https://github.com/cilium/cilium/pull/26962/commits/d288a7590c0d2c6e198457c39c7974366f823a9a#r1307537605.

File: pkg/ipam/multipool.go
Comment (danehans) [inline pkg/ipam/multipool.go]: s/nodes/node/

File: pkg/ipam/clusterpool.go
Comment (bimmlerd) [reply pkg/ipam/clusterpool.go]: clusterpool was removed

File: pkg/ipam/clusterpool.go
Comment (bimmlerd) [reply pkg/ipam/clusterpool.go]: clusterpool was removed

File: pkg/bgp/speaker/speaker.go
Comment (bimmlerd) [reply pkg/bgp/speaker/speaker.go:21]: I've chosen to be consistent with the above, `k8sRuntime` instead. We don't seem to have consistency in the codebase here, hence unclear which would be correct.

File: pkg/bgp/speaker/speaker.go
Comment (bimmlerd) [reply pkg/bgp/speaker/speaker.go]: It's the daemon context which ends up being passed in - changes such that it is passed to the subscription too.

File: pkg/ipam/multipool.go
Comment (bimmlerd) [reply pkg/ipam/multipool.go]: We do not. Previous code was getting called via the k8s informer, hence I believe this to be equivalent. Added a comment to that effect.

**Comment by bimmlerd** [comment]
Finally got around to updating this. Significant changes since last review:

1. clusterpool has been dropped upstream, hence dropping changes to it in this PR.
2. improved context handling in the speaker, and added an explanation which in IPAM there isn't a signal available to stop.

PTAL.

**Review by gandro — APPROVED** [review]
Looks good to me from my code owners! Thanks a lot for taking care of this!

**Review by rastislavs — APPROVED** [review]
LGTM

File: pkg/bgp/speaker/speaker.go
Comment (joamaki) [inline pkg/bgp/speaker/speaker.go:290]: Resource[T] doesn't log failed events (as it doesn't have enough context, so it's left to subscribers). Could you check whether how this now differs from K8sWatcher and whether logging behavior changed?

File: pkg/ipam/multipool.go
Comment (joamaki) [inline pkg/ipam/multipool.go]: nit: Perhaps call this `startCiliumNodeEventLoop` so it's clear it's not blocking and what it's handling?

File: pkg/ipam/multipool.go
Comment (joamaki) [reply pkg/ipam/multipool.go]: Thanks!

**Review by joamaki — APPROVED** [review]

**Comment by bimmlerd** [comment]
want to address Jussi's nits before merging

File: pkg/bgp/speaker/speaker.go
Comment (bimmlerd) [reply pkg/bgp/speaker/speaker.go:290]: The logging behaviour did not change AFAICT - cilium node chain did not log by itself; and the metric of k8s event success is baked into resource, so no change there.

**Comment by danehans** [comment]
@bimmlerd the ipsec upgrade job is [failing](https://github.com/cilium/cilium/actions/runs/6393490426/job/17352832845#step:16:123) due to this error:

```sh
Unable to contact Hubble Relay, disabling Hubble telescope and flow validation: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing: dial tcp: lookup localhost on 1.1.1.1:53: no such host"
```

I don't understand why Cilium Agent is using a DNS server to resolve `localhost`. It appears to be related to [this](https://github.com/cilium/cilium/issues/27799) flake.

**Comment by bimmlerd** [comment]
> @bimmlerd the ipsec upgrade job is [failing](https://github.com/cilium/cilium/actions/runs/6393490426/job/17352832845#step:16:123) due to this error:
> 
> ```shell
> Unable to contact Hubble Relay, disabling Hubble telescope and flow validation: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing: dial tcp: lookup localhost on 1.1.1.1:53: no such host"
> ```
> 
> I don't understand why Cilium Agent is using a DNS server to resolve `localhost`. It appears to be related to [this](https://github.com/cilium/cilium/issues/27799) flake.

I don't think so. AFAICT it's failing because of `  🟥 Pod test-conn-disrupt-client-54df7db847-nqmzw flow was interrupted (restart count does not match 2 != 3)`
In any case, this isn't a required test, which is typically a strong signal that the test is either still in development or was deemed too flaky. The new failure is different than the two before that, so I'm not considering it as a signal of breakage.

**Pull request merged**
